// @livekit/agents-plugin-elevenlabs/dist/stt.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var stt_exports = {};
__export(stt_exports, {
  STT: () => STT,
  SpeechStream: () => SpeechStream
});
module.exports = __toCommonJS(stt_exports);
var import_agents = require("@livekit/agents");
var import_ws = require("ws");
var import_utils = require("./_utils.cjs");
const API_BASE_URL_V1 = "https://api.elevenlabs.io/v1";
const AUTHORIZATION_HEADER = "xi-api-key";
function isAbortSignal(value) {
  return typeof value === "object" && value !== null && "aborted" in value && "addEventListener" in value;
}
function createWav(frame) {
  const bitsPerSample = 16;
  const byteRate = frame.sampleRate * frame.channels * bitsPerSample / 8;
  const blockAlign = frame.channels * bitsPerSample / 8;
  const header = Buffer.alloc(44);
  header.write("RIFF", 0);
  header.writeUInt32LE(36 + frame.data.byteLength, 4);
  header.write("WAVE", 8);
  header.write("fmt ", 12);
  header.writeUInt32LE(16, 16);
  header.writeUInt16LE(1, 20);
  header.writeUInt16LE(frame.channels, 22);
  header.writeUInt32LE(frame.sampleRate, 24);
  header.writeUInt32LE(byteRate, 28);
  header.writeUInt16LE(blockAlign, 32);
  header.writeUInt16LE(bitsPerSample, 34);
  header.write("data", 36);
  header.writeUInt32LE(frame.data.byteLength, 40);
  const pcm = Buffer.from(frame.data.buffer, frame.data.byteOffset, frame.data.byteLength);
  return Buffer.concat([header, pcm]);
}
function toRecord(value) {
  return typeof value === "object" && value !== null ? value : {};
}
function asString(value) {
  return typeof value === "string" ? value : void 0;
}
function asNumber(value) {
  return typeof value === "number" ? value : void 0;
}
function asWords(value) {
  if (!Array.isArray(value)) return [];
  return value.map((word) => {
    const record = toRecord(word);
    return {
      text: asString(record.text),
      start: asNumber(record.start),
      end: asNumber(record.end),
      speaker_id: asString(record.speaker_id) ?? null
    };
  });
}
function parseBatchResponse(value) {
  const record = toRecord(value);
  return {
    text: asString(record.text),
    language_code: asString(record.language_code),
    words: asWords(record.words),
    detail: asString(record.detail)
  };
}
function parseStreamEvent(value) {
  const record = toRecord(value);
  return {
    message_type: asString(record.message_type),
    text: asString(record.text),
    words: asWords(record.words),
    language_code: asString(record.language_code),
    session_id: asString(record.session_id),
    message: asString(record.message),
    details: asString(record.details)
  };
}
class STT extends import_agents.stt.STT {
  #opts;
  #session;
  #streams = /* @__PURE__ */ new Set();
  #logger = (0, import_agents.log)();
  label = "elevenlabs.STT";
  constructor(opts = {}) {
    let modelId = opts.modelId;
    if (opts.useRealtime !== void 0) {
      if (modelId !== void 0) {
        (0, import_agents.log)().warn(
          "both `useRealtime` and `modelId` parameters are provided. `useRealtime` will be ignored."
        );
      } else {
        (0, import_agents.log)().warn(
          "`useRealtime` parameter is deprecated. Specify a realtime modelId to enable streaming. Defaulting modelId to one based on useRealtime parameter."
        );
        modelId = opts.useRealtime ? "scribe_v2_realtime" : "scribe_v1";
      }
    }
    modelId = modelId ?? "scribe_v1";
    const useRealtime = modelId === "scribe_v2_realtime";
    if (!useRealtime && opts.serverVad !== void 0) {
      (0, import_agents.log)().warn("Server-side VAD is only supported for Scribe v2 realtime model");
    }
    const includeTimestamps = opts.includeTimestamps ?? false;
    super({
      streaming: useRealtime,
      interimResults: true,
      alignedTranscript: includeTimestamps && useRealtime ? "word" : false
    });
    const apiKey = opts.apiKey ?? process.env.ELEVEN_API_KEY;
    if (!apiKey) {
      throw new Error(
        "ElevenLabs API key is required, either as argument or set ELEVEN_API_KEY environmental variable"
      );
    }
    this.#opts = {
      apiKey,
      baseURL: opts.baseURL ?? API_BASE_URL_V1,
      languageCode: opts.languageCode ? (0, import_agents.normalizeLanguage)(opts.languageCode) : void 0,
      tagAudioEvents: opts.tagAudioEvents ?? true,
      sampleRate: opts.sampleRate ?? 16e3,
      serverVad: opts.serverVad,
      includeTimestamps,
      modelId,
      keyterms: opts.keyterms
    };
    this.#session = opts.httpSession ?? {};
  }
  get model() {
    return this.#opts.modelId;
  }
  get provider() {
    return "ElevenLabs";
  }
  async recognize(buffer, optionsOrAbortSignal) {
    const options = isAbortSignal(optionsOrAbortSignal) ? { abortSignal: optionsOrAbortSignal } : optionsOrAbortSignal ?? {};
    const connOptions = options.connOptions ?? import_agents.DEFAULT_API_CONNECT_OPTIONS;
    for (let i = 0; i < connOptions.maxRetry + 1; i++) {
      try {
        const startTime = process.hrtime.bigint();
        const event = await this.#recognizeImpl(buffer, options.language, options.abortSignal);
        const durationMs = Number((process.hrtime.bigint() - startTime) / BigInt(1e6));
        this.emit("metrics_collected", {
          type: "stt_metrics",
          requestId: event.requestId ?? "",
          timestamp: Date.now(),
          durationMs,
          label: this.label,
          audioDurationMs: Math.round((0, import_agents.calculateAudioDurationSeconds)(buffer) * 1e3),
          streamed: false,
          metadata: {
            modelProvider: this.provider,
            modelName: this.model
          }
        });
        return event;
      } catch (error) {
        if (error instanceof import_agents.APIError) {
          const retryInterval = (0, import_agents.intervalForRetry)(connOptions, i);
          if (connOptions.maxRetry === 0 || !error.retryable) {
            this.#emitError(error, false);
            throw error;
          } else if (i === connOptions.maxRetry) {
            this.#emitError(error, false);
            throw new import_agents.APIConnectionError({
              message: `failed to recognize speech after ${connOptions.maxRetry + 1} attempts`,
              options: { retryable: false }
            });
          } else {
            this.#logger.warn(
              { stt: this.label, attempt: i + 1, error },
              `failed to recognize speech, retrying in ${retryInterval}ms`
            );
          }
          if (retryInterval > 0) {
            await (0, import_agents.delay)(retryInterval);
          }
        } else {
          throw error;
        }
      }
    }
    throw new import_agents.APIConnectionError({ message: "failed to recognize speech" });
  }
  async _recognize(buffer, abortSignal) {
    return this.#recognizeImpl(buffer, void 0, abortSignal);
  }
  #emitError(error, recoverable) {
    this.emit("error", {
      type: "stt_error",
      timestamp: Date.now(),
      label: this.label,
      error,
      recoverable
    });
  }
  async #recognizeImpl(buffer, language, abortSignal) {
    var _a;
    if (language !== void 0) {
      this.#opts.languageCode = (0, import_agents.normalizeLanguage)(language);
    }
    const wavBytes = createWav((0, import_agents.mergeFrames)(buffer));
    const form = new FormData();
    form.append("file", new Blob([new Uint8Array(wavBytes)], { type: "audio/x-wav" }), "audio.wav");
    form.append("model_id", this.#opts.modelId);
    form.append("tag_audio_events", String(this.#opts.tagAudioEvents));
    if (this.#opts.languageCode) {
      form.append("language_code", this.#opts.languageCode);
    }
    if (this.#opts.keyterms !== void 0) {
      for (const keyterm of this.#opts.keyterms) {
        form.append("keyterms", keyterm);
      }
    }
    try {
      const fetchFn = this.#session.fetch ?? fetch;
      const response = await fetchFn(`${this.#opts.baseURL}/speech-to-text`, {
        method: "POST",
        headers: { [AUTHORIZATION_HEADER]: this.#opts.apiKey },
        body: form,
        signal: abortSignal ?? null
      });
      const responseJson = parseBatchResponse(await response.json());
      if (response.status !== 200) {
        throw new import_agents.APIStatusError({
          message: responseJson.detail ?? "Unknown ElevenLabs error",
          options: { statusCode: response.status, requestId: null, body: responseJson }
        });
      }
      const words = responseJson.words ?? [];
      const speakerId = words.length > 0 ? ((_a = words[0]) == null ? void 0 : _a.speaker_id) ?? null : null;
      const startTime = words.length > 0 ? Math.min(...words.map((word) => word.start ?? 0)) : 0;
      const endTime = words.length > 0 ? Math.max(...words.map((word) => word.end ?? 0)) : 0;
      const normalizedLanguage = (0, import_agents.normalizeLanguage)(
        responseJson.language_code ?? this.#opts.languageCode ?? ""
      );
      return this.#transcriptionToSpeechEvent(
        normalizedLanguage,
        responseJson.text ?? "",
        startTime,
        endTime,
        speakerId,
        words.length > 0 ? words : void 0
      );
    } catch (error) {
      if (error instanceof import_agents.APITimeoutError || error instanceof import_agents.APIConnectionError) {
        throw error;
      }
      if (error instanceof import_agents.APIStatusError) {
        throw new import_agents.APIConnectionError({ message: error.message });
      }
      if (error instanceof Error && error.name === "AbortError") {
        throw new import_agents.APITimeoutError({});
      }
      throw new import_agents.APIConnectionError({});
    }
  }
  #transcriptionToSpeechEvent(languageCode, text, startTime, endTime, speakerId, words) {
    return {
      type: import_agents.stt.SpeechEventType.FINAL_TRANSCRIPT,
      alternatives: [
        {
          text,
          language: (0, import_agents.normalizeLanguage)(languageCode),
          speakerId,
          startTime,
          endTime,
          confidence: 0,
          words: words == null ? void 0 : words.map(
            (word) => (0, import_agents.createTimedString)({
              text: word.text ?? "",
              startTime: word.start ?? 0,
              endTime: word.end ?? 0
            })
          )
        }
      ]
    };
  }
  updateOptions(opts) {
    if (opts.tagAudioEvents !== void 0) {
      this.#opts.tagAudioEvents = opts.tagAudioEvents;
    }
    if (opts.serverVad !== void 0) {
      this.#opts.serverVad = opts.serverVad;
    }
    if (opts.keyterms !== void 0) {
      this.#opts.keyterms = opts.keyterms;
    }
    for (const ref of this.#streams) {
      const stream = ref.deref();
      if (stream) {
        stream.updateOptions({ serverVad: opts.serverVad });
      } else {
        this.#streams.delete(ref);
      }
    }
  }
  stream(options) {
    const stream = new SpeechStream(
      this,
      this.#opts,
      (options == null ? void 0 : options.connOptions) ?? import_agents.DEFAULT_API_CONNECT_OPTIONS,
      (options == null ? void 0 : options.language) !== void 0 ? (0, import_agents.normalizeLanguage)(options.language) : this.#opts.languageCode,
      this.#session
    );
    this.#streams.add(new WeakRef(stream));
    return stream;
  }
}
class SpeechStream extends import_agents.stt.SpeechStream {
  #opts;
  #language;
  #session;
  #reconnectEvent = new import_agents.Future();
  #speaking = false;
  #audioDurationCollector;
  #logger = (0, import_agents.log)();
  label = "elevenlabs.SpeechStream";
  constructor(sttInstance, opts, connOptions, language, httpSession) {
    super(sttInstance, opts.sampleRate, connOptions);
    this.#opts = opts;
    this.#language = language;
    this.#session = httpSession;
    this.#audioDurationCollector = new import_utils.PeriodicCollector(
      (duration) => this.#onAudioDurationReport(duration),
      { duration: 5 }
    );
  }
  updateOptions(opts) {
    if (opts.serverVad !== void 0) {
      this.#opts.serverVad = opts.serverVad;
      if (!this.#reconnectEvent.done) {
        this.#reconnectEvent.resolve();
      }
    }
  }
  #onAudioDurationReport(duration) {
    this.queue.put({
      type: import_agents.stt.SpeechEventType.RECOGNITION_USAGE,
      recognitionUsage: { audioDuration: duration }
    });
  }
  async run() {
    while (!this.closed && !this.input.closed) {
      let ws = null;
      let closingWs = false;
      const sessionController = new AbortController();
      try {
        ws = await this.#connectWs();
        const keepaliveTask = import_agents.Task.from(async (controller) => {
          while (!controller.signal.aborted) {
            await Promise.race([(0, import_agents.delay)(1e4), (0, import_agents.waitForAbort)(controller.signal)]);
            if (controller.signal.aborted || (ws == null ? void 0 : ws.readyState) !== import_ws.WebSocket.OPEN) return;
            ws.send(
              JSON.stringify({
                message_type: "input_audio_chunk",
                audio_base_64: "",
                commit: false,
                sample_rate: this.#opts.sampleRate
              })
            );
          }
        }, sessionController);
        const sendTask = import_agents.Task.from(async (controller) => {
          const samples50Ms = Math.floor(this.#opts.sampleRate / 20);
          const audioByteStream = new import_agents.AudioByteStream(this.#opts.sampleRate, 1, samples50Ms);
          const abortPromise = (0, import_agents.waitForAbort)(controller.signal);
          const streamAbortPromise = (0, import_agents.waitForAbort)(this.abortSignal);
          let hasEnded = false;
          try {
            while (!this.closed) {
              const result = await Promise.race([
                this.input.next(),
                abortPromise,
                streamAbortPromise
              ]);
              if (result === void 0) return;
              if (result.done) break;
              const data = result.value;
              let frames;
              if (data === SpeechStream.FLUSH_SENTINEL) {
                frames = audioByteStream.flush();
                hasEnded = true;
              } else {
                frames = audioByteStream.write(data.data);
              }
              for (const frame of frames) {
                this.#audioDurationCollector.push(frame.samplesPerChannel / frame.sampleRate);
                const audioBase64 = Buffer.from(
                  frame.data.buffer,
                  frame.data.byteOffset,
                  frame.data.byteLength
                ).toString("base64");
                ws == null ? void 0 : ws.send(
                  JSON.stringify({
                    message_type: "input_audio_chunk",
                    audio_base_64: audioBase64,
                    commit: false,
                    sample_rate: this.#opts.sampleRate
                  })
                );
                if (hasEnded) {
                  this.#audioDurationCollector.flush();
                  hasEnded = false;
                }
              }
            }
          } finally {
            closingWs = true;
          }
        }, sessionController);
        const recvTask = import_agents.Task.from(async (controller) => {
          const receiveMessages = new Promise((resolve, reject) => {
            const onMessage = (msg, isBinary) => {
              if (isBinary) {
                this.#logger.warn("unexpected ElevenLabs STT binary message");
                return;
              }
              try {
                this.#processStreamEvent(parseStreamEvent(JSON.parse(msg.toString())));
              } catch (error) {
                this.#logger.error({ error }, "failed to process ElevenLabs STT message");
              }
            };
            const onClose = (code, reason) => {
              sessionController.abort();
              if (closingWs || this.closed) {
                resolve();
                return;
              }
              reject(
                new import_agents.APIStatusError({
                  message: "ElevenLabs STT connection closed unexpectedly",
                  options: {
                    statusCode: code || -1,
                    body: { reason: reason.toString() }
                  }
                })
              );
            };
            const onError = (error) => {
              sessionController.abort();
              reject(new import_agents.APIConnectionError({ message: error.message }));
            };
            ws == null ? void 0 : ws.on("message", onMessage);
            ws == null ? void 0 : ws.once("close", onClose);
            ws == null ? void 0 : ws.once("error", onError);
          });
          await Promise.race([receiveMessages, (0, import_agents.waitForAbort)(controller.signal)]);
        }, sessionController);
        const runResult = await Promise.race([
          Promise.all([keepaliveTask.result, sendTask.result, recvTask.result]).then(
            () => "done"
          ),
          this.#reconnectEvent.await.then(() => "reconnect"),
          (0, import_agents.waitForAbort)(this.abortSignal).then(() => "abort")
        ]);
        if (runResult === "reconnect") {
          this.#reconnectEvent = new import_agents.Future();
          continue;
        }
        break;
      } finally {
        closingWs = true;
        sessionController.abort();
        ws == null ? void 0 : ws.close();
      }
    }
  }
  async #connectWs() {
    const commitStrategy = this.#opts.serverVad === null ? "manual" : "vad";
    const params = [
      `model_id=${this.#opts.modelId}`,
      `audio_format=pcm_${this.#opts.sampleRate}`,
      `commit_strategy=${commitStrategy}`
    ];
    if (!this.#language) {
      params.push("include_language_detection=true");
    }
    if (this.#opts.serverVad) {
      if (this.#opts.serverVad.vadSilenceThresholdSecs !== void 0 && this.#opts.serverVad.vadSilenceThresholdSecs !== null) {
        params.push(`vad_silence_threshold_secs=${this.#opts.serverVad.vadSilenceThresholdSecs}`);
      }
      if (this.#opts.serverVad.vadThreshold !== void 0 && this.#opts.serverVad.vadThreshold !== null) {
        params.push(`vad_threshold=${this.#opts.serverVad.vadThreshold}`);
      }
      if (this.#opts.serverVad.minSpeechDurationMs !== void 0 && this.#opts.serverVad.minSpeechDurationMs !== null) {
        params.push(`min_speech_duration_ms=${this.#opts.serverVad.minSpeechDurationMs}`);
      }
      if (this.#opts.serverVad.minSilenceDurationMs !== void 0 && this.#opts.serverVad.minSilenceDurationMs !== null) {
        params.push(`min_silence_duration_ms=${this.#opts.serverVad.minSilenceDurationMs}`);
      }
    }
    if (this.#language) {
      params.push(`language_code=${this.#language}`);
    }
    if (this.#opts.includeTimestamps) {
      params.push("include_timestamps=true");
    }
    const baseURL = this.#opts.baseURL.replace("https://", "wss://").replace("http://", "ws://");
    const wsUrl = `${baseURL}/speech-to-text/realtime?${params.join("&")}`;
    const headers = { [AUTHORIZATION_HEADER]: this.#opts.apiKey };
    try {
      if (this.#session.wsConnect) {
        return await this.#session.wsConnect(wsUrl, {
          headers,
          signal: this.abortSignal,
          timeoutMs: import_agents.DEFAULT_API_CONNECT_OPTIONS.timeoutMs
        });
      }
      const ws = new import_ws.WebSocket(wsUrl, { headers });
      await new Promise((resolve, reject) => {
        const timeout = setTimeout(() => {
          ws.close();
          reject(new Error("connect timeout"));
        }, import_agents.DEFAULT_API_CONNECT_OPTIONS.timeoutMs);
        const cleanup = () => {
          clearTimeout(timeout);
          this.abortSignal.removeEventListener("abort", onAbort);
          ws.off("open", onOpen);
          ws.off("error", onError);
          ws.off("close", onClose);
        };
        const onOpen = () => {
          cleanup();
          resolve();
        };
        const onError = (error) => {
          cleanup();
          reject(error);
        };
        const onClose = (code) => {
          cleanup();
          reject(new Error(`WebSocket returned ${code}`));
        };
        const onAbort = () => {
          cleanup();
          ws.close();
          reject(new Error("aborted"));
        };
        ws.once("open", onOpen);
        ws.once("error", onError);
        ws.once("close", onClose);
        this.abortSignal.addEventListener("abort", onAbort, { once: true });
      });
      return ws;
    } catch (error) {
      throw new import_agents.APIConnectionError({ message: "Failed to connect to ElevenLabs" });
    }
  }
  #processStreamEvent(data) {
    var _a, _b;
    const messageType = data.message_type;
    const text = data.text ?? "";
    const words = data.words ?? [];
    const startTime = words.length > 0 ? ((_a = words[0]) == null ? void 0 : _a.start) ?? 0 : 0;
    const endTime = words.length > 0 ? ((_b = words[words.length - 1]) == null ? void 0 : _b.end) ?? 0 : 0;
    const languageCode = data.language_code ?? this.#language;
    const normalizedLanguage = languageCode ? (0, import_agents.normalizeLanguage)(languageCode) : (0, import_agents.normalizeLanguage)("en");
    const speechData = {
      language: normalizedLanguage,
      text,
      startTime: startTime + this.startTimeOffset,
      endTime: endTime + this.startTimeOffset,
      confidence: 0
    };
    if (words.length > 0) {
      speechData.words = words.map(
        (word) => (0, import_agents.createTimedString)({
          text: word.text ?? "",
          startTime: (word.start ?? 0) + this.startTimeOffset,
          endTime: (word.end ?? 0) + this.startTimeOffset,
          startTimeOffset: this.startTimeOffset
        })
      );
    }
    if (messageType === "partial_transcript") {
      this.#logger.debug({ data }, "Received message type partial_transcript");
      if (text) {
        if (!this.#speaking) {
          this.queue.put({ type: import_agents.stt.SpeechEventType.START_OF_SPEECH });
          this.#speaking = true;
        }
        this.queue.put({
          type: import_agents.stt.SpeechEventType.INTERIM_TRANSCRIPT,
          alternatives: [speechData]
        });
      }
    } else if (messageType === "committed_transcript" && !this.#opts.includeTimestamps || messageType === "committed_transcript_with_timestamps" && this.#opts.includeTimestamps) {
      if (text) {
        if (!this.#speaking) {
          this.queue.put({ type: import_agents.stt.SpeechEventType.START_OF_SPEECH });
          this.#speaking = true;
        }
        this.queue.put({
          type: import_agents.stt.SpeechEventType.FINAL_TRANSCRIPT,
          alternatives: [speechData]
        });
      } else if (this.#speaking) {
        this.queue.put({ type: import_agents.stt.SpeechEventType.END_OF_SPEECH });
        this.#speaking = false;
      }
    } else if (messageType === "committed_transcript") {
      return;
    } else if (messageType === "session_started") {
      this.#logger.debug(`Session started with ID: ${data.session_id ?? "unknown"}`);
    } else if (messageType === "auth_error" || messageType === "quota_exceeded" || messageType === "transcriber_error" || messageType === "input_error" || messageType === "error") {
      const errorMsg = data.message ?? "Unknown error";
      const detailsSuffix = data.details ? ` - ${data.details}` : "";
      this.#logger.error(`ElevenLabs STT error [${messageType}]: ${errorMsg}${detailsSuffix}`);
      throw new import_agents.APIConnectionError({ message: `${messageType}: ${errorMsg}${detailsSuffix}` });
    } else if (messageType === "committed_transcript_with_timestamps" && !this.#opts.includeTimestamps) {
      return;
    } else {
      this.#logger.warn(`ElevenLabs STT unknown message type: ${messageType}, data: ${data}`);
    }
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  STT,
  SpeechStream
});
