// @livekit/agents-plugin-elevenlabs/dist/tts.cjs
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var tts_exports = {};
__export(tts_exports, {
  ChunkedStream: () => ChunkedStream,
  SynthesizeStream: () => SynthesizeStream,
  TTS: () => TTS
});
module.exports = __toCommonJS(tts_exports);
var import_agents = require("@livekit/agents");
var import_mutex = require("@livekit/mutex");
var import_ws = require("ws");
const DEFAULT_VOICE_ID = "bIHbv24MWmeRgasZH58o";
const API_BASE_URL_V1 = "https://api.elevenlabs.io/v1";
const AUTHORIZATION_HEADER = "xi-api-key";
const WS_INACTIVITY_TIMEOUT = 180;
const DEFAULT_ENCODING = "pcm_22050";
function sampleRateFromFormat(encoding) {
  const split = encoding.split("_");
  return parseInt(split[1], 10);
}
function synthesizeUrl(opts) {
  const { baseURL, voiceId, model, encoding, streamingLatency } = opts;
  let url = `${baseURL}/text-to-speech/${voiceId}/stream?model_id=${model}&output_format=${encoding}`;
  if (streamingLatency !== void 0) {
    url += `&optimize_streaming_latency=${streamingLatency}`;
  }
  return url;
}
function multiStreamUrl(opts) {
  const baseURL = opts.baseURL.replace("https://", "wss://").replace("http://", "ws://");
  const params = [];
  params.push(`model_id=${opts.model}`);
  params.push(`output_format=${opts.encoding}`);
  if (opts.language) {
    params.push(`language_code=${(0, import_agents.getBaseLanguage)(opts.language)}`);
  }
  params.push(`enable_ssml_parsing=${opts.enableSsmlParsing}`);
  params.push(`enable_logging=${opts.enableLogging}`);
  params.push(`inactivity_timeout=${opts.inactivityTimeout}`);
  params.push(`apply_text_normalization=${opts.applyTextNormalization}`);
  if (opts.applyLanguageTextNormalization !== void 0) {
    params.push(`apply_language_text_normalization=${opts.applyLanguageTextNormalization}`);
  }
  if (opts.syncAlignment) {
    params.push("sync_alignment=true");
  }
  if (opts.autoMode !== void 0) {
    params.push(`auto_mode=${opts.autoMode}`);
  }
  return `${baseURL}/text-to-speech/${opts.voiceId}/multi-stream-input?${params.join("&")}`;
}
function stripUndefined(obj) {
  const result = {};
  for (const key in obj) {
    if (obj[key] !== void 0) {
      result[key] = obj[key];
    }
  }
  return result;
}
function toTimedWords(text, startTimesMs, durationsMs, flush = false, firstWordOffsetMs = 0) {
  if (!text || startTimesMs.length === 0 || durationsMs.length === 0) {
    return [[], text || ""];
  }
  const lastStartTime = startTimesMs[startTimesMs.length - 1];
  const lastDuration = durationsMs[durationsMs.length - 1];
  const timestamps = [...startTimesMs, lastStartTime + lastDuration];
  const words = import_agents.tokenize.basic.splitWords(text, false);
  const timedWords = [];
  if (words.length === 0) {
    return [[], text];
  }
  const startIndices = words.map((w) => w[1]);
  let end = 0;
  for (let i = 0; i < startIndices.length - 1; i++) {
    const start = startIndices[i];
    const nextStart = startIndices[i + 1];
    end = nextStart;
    const startT = Math.max(0, (timestamps[start] ?? 0) - firstWordOffsetMs) / 1e3;
    const endT = Math.max(0, (timestamps[nextStart] ?? 0) - firstWordOffsetMs) / 1e3;
    timedWords.push(
      (0, import_agents.createTimedString)({
        text: text.slice(start, nextStart),
        startTime: startT,
        endTime: endT
      })
    );
  }
  if (flush && words.length > 0) {
    const lastWordStart = startIndices[startIndices.length - 1];
    const startT = Math.max(0, (timestamps[lastWordStart] ?? 0) - firstWordOffsetMs) / 1e3;
    const endT = Math.max(0, (timestamps[timestamps.length - 1] ?? 0) - firstWordOffsetMs) / 1e3;
    timedWords.push(
      (0, import_agents.createTimedString)({
        text: text.slice(lastWordStart),
        startTime: startT,
        endTime: endT
      })
    );
    end = text.length;
  } else if (words.length > 0) {
    end = startIndices[startIndices.length - 1];
  }
  return [timedWords, text.slice(end)];
}
class Connection {
  #opts;
  #ws = null;
  #isCurrent = true;
  #activeContexts = /* @__PURE__ */ new Set();
  #inputQueue = [];
  #contextData = /* @__PURE__ */ new Map();
  #sendTask = null;
  #recvTask = null;
  #closed = false;
  #logger = (0, import_agents.log)();
  #inputQueueResolver = null;
  constructor(opts) {
    this.#opts = opts;
  }
  get voiceId() {
    return this.#opts.voiceId;
  }
  get isCurrent() {
    return this.#isCurrent;
  }
  get closed() {
    return this.#closed;
  }
  markNonCurrent() {
    this.#isCurrent = false;
  }
  async connect() {
    if (this.#ws || this.#closed) {
      return;
    }
    const url = multiStreamUrl(this.#opts);
    const headers = { [AUTHORIZATION_HEADER]: this.#opts.apiKey };
    return new Promise((resolve, reject) => {
      this.#ws = new import_ws.WebSocket(url, { headers });
      this.#ws.on("open", () => {
        this.#sendTask = this.#sendLoop();
        this.#recvTask = this.#recvLoop();
        resolve();
      });
      this.#ws.on("error", (error) => {
        this.#logger.error({ error }, "WebSocket connection error");
        reject(new import_agents.APIConnectionError({ message: `WebSocket error: ${error.message}` }));
      });
    });
  }
  registerStream(stream2, waiter) {
    const contextId = stream2.contextId;
    this.#contextData.set(contextId, {
      stream: stream2,
      waiter,
      textBuffer: "",
      startTimesMs: [],
      durationsMs: [],
      firstWordOffsetMs: null
    });
  }
  sendContent(content) {
    var _a;
    if (this.#closed || !this.#ws || this.#ws.readyState !== import_ws.WebSocket.OPEN) {
      throw new import_agents.APIConnectionError({ message: "WebSocket connection is closed" });
    }
    this.#inputQueue.push(content);
    (_a = this.#inputQueueResolver) == null ? void 0 : _a.call(this);
  }
  closeContext(contextId) {
    var _a;
    if (this.#closed || !this.#ws || this.#ws.readyState !== import_ws.WebSocket.OPEN) {
      throw new import_agents.APIConnectionError({ message: "WebSocket connection is closed" });
    }
    this.#inputQueue.push({ contextId });
    (_a = this.#inputQueueResolver) == null ? void 0 : _a.call(this);
  }
  async #sendLoop() {
    try {
      while (!this.#closed) {
        if (this.#inputQueue.length === 0) {
          await new Promise((resolve) => {
            this.#inputQueueResolver = resolve;
          });
          this.#inputQueueResolver = null;
        }
        if (this.#closed) break;
        const msg = this.#inputQueue.shift();
        if (!msg) continue;
        if (!this.#ws || this.#ws.readyState !== import_ws.WebSocket.OPEN) {
          break;
        }
        if ("text" in msg) {
          const content = msg;
          const isNewContext = !this.#activeContexts.has(content.contextId);
          if (!this.#isCurrent && isNewContext) {
            continue;
          }
          if (isNewContext) {
            const voiceSettings = this.#opts.voiceSettings ? stripUndefined(this.#opts.voiceSettings) : {};
            const initPkt = {
              text: " ",
              voice_settings: voiceSettings,
              context_id: content.contextId
            };
            if (this.#opts.chunkLengthSchedule) {
              initPkt.generation_config = {
                chunk_length_schedule: this.#opts.chunkLengthSchedule
              };
            }
            if (this.#opts.pronunciationDictionaryLocators) {
              initPkt.pronunciation_dictionary_locators = this.#opts.pronunciationDictionaryLocators.map((locator) => ({
                pronunciation_dictionary_id: locator.pronunciation_dictionary_id,
                version_id: locator.version_id
              }));
            }
            const initPktStr = JSON.stringify(initPkt);
            this.#ws.send(initPktStr);
            this.#activeContexts.add(content.contextId);
          }
          const pkt = {
            text: content.text,
            context_id: content.contextId
          };
          if (content.flush) {
            pkt.flush = true;
          }
          const pktStr = JSON.stringify(pkt);
          this.#ws.send(pktStr);
        } else {
          const closeMsg = msg;
          if (this.#activeContexts.has(closeMsg.contextId)) {
            const closePkt = {
              context_id: closeMsg.contextId,
              close_context: true
            };
            const closePktStr = JSON.stringify(closePkt);
            this.#ws.send(closePktStr);
          }
        }
      }
    } catch (e) {
      this.#logger.warn({ error: e }, "send loop error");
    } finally {
      if (!this.#closed) {
        await this.close();
      }
    }
  }
  async #recvLoop() {
    try {
      const messageChannel = import_agents.stream.createStreamChannel();
      const errorFuture = new import_agents.Future();
      const onMessage = (rawData) => {
        try {
          const parsed = JSON.parse(rawData.toString());
          messageChannel.write(parsed);
        } catch (e) {
          this.#logger.warn({ error: e }, "failed to parse WebSocket message");
        }
      };
      const onClose = (code) => {
        if (!this.#closed && this.#contextData.size > 0) {
          errorFuture.reject(
            new import_agents.APIStatusError({
              message: "ElevenLabs websocket connection closed unexpectedly",
              options: { statusCode: code }
            })
          );
        }
        messageChannel.close();
      };
      const onError = (error) => {
        errorFuture.reject(error);
        messageChannel.close();
      };
      if (!this.#ws) return;
      this.#ws.on("message", onMessage);
      this.#ws.on("close", onClose);
      this.#ws.on("error", onError);
      const cleanup = () => {
        var _a, _b, _c;
        (_a = this.#ws) == null ? void 0 : _a.off("message", onMessage);
        (_b = this.#ws) == null ? void 0 : _b.off("close", onClose);
        (_c = this.#ws) == null ? void 0 : _c.off("error", onError);
      };
      const reader = messageChannel.stream().getReader();
      try {
        while (!this.#closed) {
          const result = await reader.read();
          if (result.done || this.#closed) break;
          const data = result.value;
          const contextId = data.contextId;
          const ctx = contextId ? this.#contextData.get(contextId) : void 0;
          if (data.error) {
            this.#logger.error(
              { context_id: contextId, error: data.error, data },
              "elevenlabs tts returned error"
            );
            if (contextId) {
              if (ctx) {
                ctx.waiter.reject(new import_agents.APIError(data.error));
              }
              this.#cleanupContext(contextId);
            }
            continue;
          }
          if (!ctx) {
            this.#logger.warn({ data }, "unexpected message received from elevenlabs tts");
            continue;
          }
          const stream2 = ctx.stream;
          const alignment = this.#opts.preferredAlignment === "normalized" ? data.normalizedAlignment : data.alignment;
          if (alignment && stream2) {
            const chars = alignment.chars;
            const starts = alignment.charStartTimesMs || alignment.charsStartTimesMs;
            const durs = alignment.charDurationsMs || alignment.charsDurationsMs;
            if (chars && starts && durs && chars.length === durs.length && starts.length === durs.length) {
              ctx.textBuffer += chars.join("");
              for (let i = 0; i < chars.length; i++) {
                const char = chars[i];
                const start = starts[i];
                const dur = durs[i];
                if (ctx.firstWordOffsetMs === null && start > 0) {
                  ctx.firstWordOffsetMs = start;
                }
                if (char.length > 1) {
                  for (let j = 0; j < char.length - 1; j++) {
                    ctx.startTimesMs.push(start);
                    ctx.durationsMs.push(0);
                  }
                }
                ctx.startTimesMs.push(start);
                ctx.durationsMs.push(dur);
              }
              const [timedWords, remainingText] = toTimedWords(
                ctx.textBuffer,
                ctx.startTimesMs,
                ctx.durationsMs,
                false,
                ctx.firstWordOffsetMs ?? 0
              );
              if (timedWords.length > 0) {
                stream2.pushTimedTranscript(timedWords);
              }
              ctx.textBuffer = remainingText;
              ctx.startTimesMs = ctx.startTimesMs.slice(-remainingText.length);
              ctx.durationsMs = ctx.durationsMs.slice(-remainingText.length);
            }
          }
          if (data.audio) {
            const audioData = Buffer.from(data.audio, "base64");
            stream2.pushAudio(audioData);
          }
          if (data.isFinal) {
            if (ctx.textBuffer) {
              const [timedWords] = toTimedWords(
                ctx.textBuffer,
                ctx.startTimesMs,
                ctx.durationsMs,
                true,
                ctx.firstWordOffsetMs ?? 0
              );
              if (timedWords.length > 0) {
                stream2.pushTimedTranscript(timedWords);
              }
            }
            stream2.markDone();
            ctx.waiter.resolve();
            this.#cleanupContext(contextId);
            if (!this.#isCurrent && this.#activeContexts.size === 0) {
              this.#logger.debug("no active contexts, shutting down connection");
              break;
            }
          }
        }
        if (errorFuture.done) {
          await errorFuture.await;
        }
      } finally {
        reader.releaseLock();
        cleanup();
      }
    } catch (e) {
      this.#logger.warn({ error: e }, "recv loop error");
      for (const ctx of this.#contextData.values()) {
        ctx.waiter.reject((0, import_agents.asError)(e));
      }
      this.#contextData.clear();
    } finally {
      if (!this.#closed) {
        await this.close();
      }
    }
  }
  #cleanupContext(contextId) {
    this.#contextData.delete(contextId);
    this.#activeContexts.delete(contextId);
  }
  async close() {
    var _a;
    if (this.#closed) {
      return;
    }
    this.#closed = true;
    (_a = this.#inputQueueResolver) == null ? void 0 : _a.call(this);
    for (const ctx of this.#contextData.values()) {
      ctx.waiter.reject(new import_agents.APIStatusError({ message: "connection closed" }));
    }
    this.#contextData.clear();
    if (this.#ws) {
      this.#ws.close();
      this.#ws = null;
    }
    if (this.#sendTask) {
      await this.#sendTask.catch(() => {
      });
    }
    if (this.#recvTask) {
      await this.#recvTask.catch(() => {
      });
    }
  }
}
class TTS extends import_agents.tts.TTS {
  #opts;
  #streams = /* @__PURE__ */ new Set();
  #currentConnection = null;
  #connectionLock = new import_mutex.Mutex();
  #logger = (0, import_agents.log)();
  label = "elevenlabs.TTS";
  constructor(opts = {}) {
    var _a, _b;
    const autoMode = opts.autoMode ?? opts.chunkLengthSchedule === void 0;
    const encoding = opts.encoding ?? DEFAULT_ENCODING;
    const sampleRate = sampleRateFromFormat(encoding);
    const syncAlignment = opts.syncAlignment ?? true;
    super(sampleRate, 1, {
      streaming: true,
      alignedTranscript: syncAlignment
    });
    const apiKey = opts.apiKey ?? process.env.ELEVEN_API_KEY;
    if (!apiKey) {
      throw new Error(
        "ElevenLabs API key is required, either as argument or set ELEVEN_API_KEY environmental variable"
      );
    }
    let wordTokenizer = opts.wordTokenizer;
    if (!wordTokenizer) {
      wordTokenizer = autoMode ? new import_agents.tokenize.basic.SentenceTokenizer() : new import_agents.tokenize.basic.WordTokenizer(false);
    } else if (autoMode && !(wordTokenizer instanceof import_agents.tokenize.SentenceTokenizer)) {
      this.#logger.warn(
        "autoMode is enabled, it expects full sentences or phrases, please provide a SentenceTokenizer instead of a WordTokenizer."
      );
    }
    const voiceId = opts.voiceId ?? ((_a = opts.voice) == null ? void 0 : _a.id) ?? DEFAULT_VOICE_ID;
    const voiceSettings = opts.voiceSettings ?? ((_b = opts.voice) == null ? void 0 : _b.settings);
    const model = opts.model ?? opts.modelID ?? "eleven_turbo_v2_5";
    const rawLanguage = opts.language ?? opts.languageCode;
    const language = rawLanguage ? (0, import_agents.normalizeLanguage)(rawLanguage) : void 0;
    this.#opts = {
      apiKey,
      voiceId,
      voiceSettings,
      model,
      language,
      baseURL: opts.baseURL ?? API_BASE_URL_V1,
      encoding,
      sampleRate,
      streamingLatency: opts.streamingLatency,
      wordTokenizer,
      chunkLengthSchedule: opts.chunkLengthSchedule,
      enableSsmlParsing: opts.enableSsmlParsing ?? false,
      enableLogging: opts.enableLogging ?? true,
      inactivityTimeout: opts.inactivityTimeout ?? WS_INACTIVITY_TIMEOUT,
      syncAlignment: opts.syncAlignment ?? true,
      applyTextNormalization: opts.applyTextNormalization ?? "auto",
      applyLanguageTextNormalization: opts.applyLanguageTextNormalization,
      preferredAlignment: opts.preferredAlignment ?? "normalized",
      autoMode,
      pronunciationDictionaryLocators: opts.pronunciationDictionaryLocators
    };
  }
  get model() {
    return this.#opts.model;
  }
  get provider() {
    return "ElevenLabs";
  }
  async listVoices() {
    const response = await fetch(`${this.#opts.baseURL}/voices`, {
      headers: { [AUTHORIZATION_HEADER]: this.#opts.apiKey }
    });
    const data = await response.json();
    return data.voices.map((v) => ({
      id: v.voice_id,
      name: v.name,
      category: v.category
    }));
  }
  updateOptions(opts) {
    let changed = false;
    if (opts.model !== void 0 && opts.model !== this.#opts.model) {
      this.#opts.model = opts.model;
      changed = true;
    }
    if (opts.voiceId !== void 0 && opts.voiceId !== this.#opts.voiceId) {
      this.#opts.voiceId = opts.voiceId;
      changed = true;
    }
    if (opts.voiceSettings !== void 0) {
      this.#opts.voiceSettings = opts.voiceSettings;
      changed = true;
    }
    if (opts.language !== void 0 && opts.language !== this.#opts.language) {
      this.#opts.language = (0, import_agents.normalizeLanguage)(opts.language);
      changed = true;
    }
    if (opts.pronunciationDictionaryLocators !== void 0) {
      this.#opts.pronunciationDictionaryLocators = opts.pronunciationDictionaryLocators;
      changed = true;
    }
    if (changed && this.#currentConnection) {
      this.#currentConnection.markNonCurrent();
      this.#currentConnection = null;
    }
  }
  async currentConnection() {
    const unlock = await this.#connectionLock.lock();
    try {
      if (this.#currentConnection && this.#currentConnection.isCurrent && !this.#currentConnection.closed) {
        return this.#currentConnection;
      }
      const conn = new Connection({ ...this.#opts });
      await conn.connect();
      this.#currentConnection = conn;
      return conn;
    } finally {
      unlock();
    }
  }
  synthesize(text) {
    return new ChunkedStream(this, text, { ...this.#opts });
  }
  stream(options) {
    const stream2 = new SynthesizeStream(this, { ...this.#opts }, options == null ? void 0 : options.connOptions);
    this.#streams.add(stream2);
    return stream2;
  }
  async close() {
    for (const stream2 of this.#streams) {
      stream2.close();
    }
    this.#streams.clear();
    if (this.#currentConnection) {
      await this.#currentConnection.close();
      this.#currentConnection = null;
    }
  }
}
class ChunkedStream extends import_agents.tts.ChunkedStream {
  #tts;
  #opts;
  #logger = (0, import_agents.log)();
  label = "elevenlabs.ChunkedStream";
  constructor(tts2, text, opts) {
    super(text, tts2);
    this.#tts = tts2;
    this.#opts = opts;
  }
  async run() {
    var _a;
    const voiceSettings = this.#opts.voiceSettings ? stripUndefined(this.#opts.voiceSettings) : void 0;
    const requestId = (0, import_agents.shortuuid)();
    const bstream = new import_agents.AudioByteStream(this.#opts.sampleRate, 1);
    try {
      const response = await fetch(synthesizeUrl(this.#opts), {
        method: "POST",
        headers: {
          [AUTHORIZATION_HEADER]: this.#opts.apiKey,
          "Content-Type": "application/json"
        },
        body: JSON.stringify({
          text: this.inputText,
          model_id: this.#opts.model,
          voice_settings: voiceSettings
        }),
        signal: this.abortSignal
      });
      if (!response.ok) {
        const errorText = await response.text();
        throw new import_agents.APIStatusError({
          message: `ElevenLabs API error: ${errorText}`,
          options: { statusCode: response.status }
        });
      }
      const contentType = response.headers.get("content-type") || "";
      if (!contentType.startsWith("audio/")) {
        const content = await response.text();
        throw new import_agents.APIError(`ElevenLabs returned non-audio data: ${content}`);
      }
      const reader = (_a = response.body) == null ? void 0 : _a.getReader();
      if (!reader) {
        throw new import_agents.APIError("No response body");
      }
      let lastFrame;
      while (true) {
        const { done, value } = await reader.read();
        if (done) break;
        for (const frame of bstream.write(value.buffer)) {
          if (lastFrame) {
            this.queue.put({ requestId, segmentId: requestId, frame: lastFrame, final: false });
          }
          lastFrame = frame;
        }
      }
      for (const frame of bstream.flush()) {
        if (lastFrame) {
          this.queue.put({ requestId, segmentId: requestId, frame: lastFrame, final: false });
        }
        lastFrame = frame;
      }
      if (lastFrame) {
        this.queue.put({ requestId, segmentId: requestId, frame: lastFrame, final: true });
      }
    } catch (e) {
      if (e instanceof import_agents.APIError) {
        throw e;
      }
      if (e instanceof Error && e.name === "AbortError") {
        return;
      }
      throw new import_agents.APIConnectionError({ message: `Connection error: ${e}` });
    }
  }
}
class SynthesizeStream extends import_agents.tts.SynthesizeStream {
  #tts;
  #opts;
  #contextId;
  #sentTokenizerStream;
  #logger = (0, import_agents.log)();
  #audioQueue = [];
  #timedTranscriptQueue = [];
  #streamDone = false;
  label = "elevenlabs.SynthesizeStream";
  constructor(tts2, opts, connOptions) {
    super(tts2, connOptions);
    this.#tts = tts2;
    this.#opts = opts;
    this.#contextId = (0, import_agents.shortuuid)();
    this.#sentTokenizerStream = this.#opts.wordTokenizer.stream();
  }
  get contextId() {
    return this.#contextId;
  }
  pushAudio(data) {
    if (this.closed || this.abortController.signal.aborted) {
      return;
    }
    this.#audioQueue.push(data);
  }
  pushTimedTranscript(timedWords) {
    this.#timedTranscriptQueue.push(...timedWords);
  }
  markDone() {
    this.#streamDone = true;
  }
  async run() {
    const requestId = this.#contextId;
    const segmentId = this.#contextId;
    const bstream = new import_agents.AudioByteStream(this.#opts.sampleRate, 1);
    let connection;
    try {
      connection = await this.#tts.currentConnection();
    } catch (e) {
      throw new import_agents.APIConnectionError({ message: "could not connect to ElevenLabs" });
    }
    let waiterReject;
    const waiterPromise = new Promise((resolve, reject) => {
      waiterReject = reject;
      connection.registerStream(this, { resolve, reject });
    });
    const abortHandler = () => {
      if (waiterReject) {
        waiterReject(new Error("Stream aborted"));
      }
    };
    this.abortController.signal.addEventListener("abort", abortHandler, { once: true });
    const inputTask = async () => {
      for await (const data of this.input) {
        if (this.abortController.signal.aborted) break;
        if (data === SynthesizeStream.FLUSH_SENTINEL) {
          this.#sentTokenizerStream.flush();
          continue;
        }
        this.#sentTokenizerStream.pushText(data);
      }
      this.#sentTokenizerStream.endInput();
    };
    const sentenceStreamTask = async () => {
      const flushOnChunk = this.#opts.wordTokenizer instanceof import_agents.tokenize.SentenceTokenizer && this.#opts.autoMode;
      let xmlContent = [];
      for await (const data of this.#sentTokenizerStream) {
        if (this.abortController.signal.aborted) break;
        let text = data.token;
        const xmlStartTokens = ["<phoneme", "<break"];
        const xmlEndTokens = ["</phoneme>", "/>"];
        if (this.#opts.enableSsmlParsing && xmlStartTokens.some((start) => text.startsWith(start)) || xmlContent.length > 0) {
          xmlContent.push(text);
          if (xmlEndTokens.some((end) => text.includes(end))) {
            text = xmlContent.join(" ");
            xmlContent = [];
          } else {
            continue;
          }
        }
        const formattedText = `${text} `;
        connection.sendContent({
          contextId: this.#contextId,
          text: formattedText,
          flush: flushOnChunk
        });
      }
      if (xmlContent.length > 0) {
        this.#logger.warn("ElevenLabs stream ended with incomplete xml content");
      }
      connection.sendContent({ contextId: this.#contextId, text: "", flush: true });
      connection.closeContext(this.#contextId);
    };
    const audioProcessTask = async () => {
      let lastFrame;
      let pendingTimedTranscripts = [];
      const sendLastFrame = (final) => {
        if (lastFrame) {
          this.queue.put({
            requestId,
            segmentId,
            frame: lastFrame,
            final,
            timedTranscripts: pendingTimedTranscripts.length > 0 ? pendingTimedTranscripts : void 0
          });
          lastFrame = void 0;
          pendingTimedTranscripts = [];
        }
      };
      while (!this.abortController.signal.aborted) {
        while (this.#timedTranscriptQueue.length > 0) {
          pendingTimedTranscripts.push(this.#timedTranscriptQueue.shift());
        }
        while (this.#audioQueue.length > 0) {
          const audioData = this.#audioQueue.shift();
          for (const frame of bstream.write(audioData.buffer)) {
            sendLastFrame(false);
            lastFrame = frame;
          }
        }
        if (this.#streamDone && this.#audioQueue.length === 0) {
          break;
        }
        await new Promise((resolve) => setTimeout(resolve, 10));
      }
      while (this.#timedTranscriptQueue.length > 0) {
        pendingTimedTranscripts.push(this.#timedTranscriptQueue.shift());
      }
      for (const frame of bstream.flush()) {
        sendLastFrame(false);
        lastFrame = frame;
      }
      sendLastFrame(true);
    };
    try {
      await Promise.all([inputTask(), sentenceStreamTask(), audioProcessTask(), waiterPromise]);
    } catch (e) {
      if (this.abortController.signal.aborted) {
        return;
      }
      if (e instanceof import_agents.APITimeoutError) {
        throw e;
      }
      if (e instanceof import_agents.APIStatusError) {
        throw e;
      }
      throw new import_agents.APIStatusError({ message: "Could not synthesize" });
    } finally {
      this.abortController.signal.removeEventListener("abort", abortHandler);
    }
  }
  close() {
    this.#audioQueue.length = 0;
    this.#timedTranscriptQueue.length = 0;
    this.#streamDone = true;
    this.#sentTokenizerStream.close();
    super.close();
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ChunkedStream,
  SynthesizeStream,
  TTS
});
