// @livekit/rtc-node/dist/audio_stream.js
import { AudioStreamType, NewAudioStreamRequest } from "@livekit/rtc-ffi-bindings";
import { AudioFrame } from "./audio_frame.js";
import { FfiClient, FfiClientEvent, FfiHandle } from "./ffi_client.js";
import { isFrameProcessor } from "./frame_processor.js";
import { log } from "./log.js";
class AudioStreamSource {
  constructor(track, sampleRateOrOptions, numChannels) {
    this.disposed = false;
    this.onEvent = (ev) => {
      var _a;
      if (!this.controller) {
        throw new Error("Stream controller not initialized");
      }
      if (ev.message.case != "audioStreamEvent" || ev.message.value.streamHandle != this.ffiHandle.handle) {
        return;
      }
      const streamEvent = ev.message.value.message;
      switch (streamEvent.case) {
        case "frameReceived":
          let frame = AudioFrame.fromOwnedInfo(streamEvent.value.frame);
          if (this.frameProcessor && this.frameProcessor.isEnabled()) {
            try {
              frame = this.frameProcessor.process(frame);
            } catch (err) {
              log.warn(`Frame processing failed, passing through original frame: ${err}`);
            }
          }
          this.controller.enqueue(frame);
          break;
        case "eos":
          FfiClient.instance.off(FfiClientEvent.FfiEvent, this.onEvent);
          this.controller.close();
          if (!this.disposed) {
            this.disposed = true;
            this.ffiHandle.dispose();
            (_a = this.frameProcessor) == null ? void 0 : _a.close();
          }
          break;
      }
    };
    if (sampleRateOrOptions !== void 0 && typeof sampleRateOrOptions !== "number") {
      this.sampleRate = sampleRateOrOptions.sampleRate ?? 48e3;
      this.numChannels = sampleRateOrOptions.numChannels ?? 1;
      if (isFrameProcessor(sampleRateOrOptions.noiseCancellation)) {
        this.frameProcessor = sampleRateOrOptions.noiseCancellation;
      } else {
        this.legacyNcOptions = sampleRateOrOptions.noiseCancellation;
      }
      this.frameSizeMs = sampleRateOrOptions.frameSizeMs;
    } else {
      this.sampleRate = sampleRateOrOptions ?? 48e3;
      this.numChannels = numChannels ?? 1;
    }
    const req = new NewAudioStreamRequest({
      type: AudioStreamType.AUDIO_STREAM_NATIVE,
      trackHandle: track.ffi_handle.handle,
      sampleRate: this.sampleRate,
      numChannels: this.numChannels,
      frameSizeMs: this.frameSizeMs,
      ...this.legacyNcOptions ? {
        audioFilterModuleId: this.legacyNcOptions.moduleId,
        audioFilterOptions: JSON.stringify(this.legacyNcOptions.options)
      } : {}
    });
    const res = FfiClient.instance.request({
      message: {
        case: "newAudioStream",
        value: req
      }
    });
    this.ffiHandle = new FfiHandle(res.stream.handle.id);
    FfiClient.instance.on(FfiClientEvent.FfiEvent, this.onEvent);
  }
  start(controller) {
    this.controller = controller;
  }
  cancel() {
    var _a;
    FfiClient.instance.off(FfiClientEvent.FfiEvent, this.onEvent);
    if (!this.disposed) {
      this.disposed = true;
      this.ffiHandle.dispose();
      (_a = this.frameProcessor) == null ? void 0 : _a.close();
    }
  }
}
class AudioStream extends ReadableStream {
  constructor(track, sampleRateOrOptions, numChannels) {
    super(new AudioStreamSource(track, sampleRateOrOptions, numChannels));
  }
}
export {
  AudioStream
};
